ipld/read.go

```diff
@@ -18,190 +16,131 @@ import (
-	// just downloading the first half.
-	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
-			if err != nil {
-				return nil, err
-			}
+	dataSquare := make([][]byte, edsWidth*edsWidth)
 
-			go sc.retrieveShare(rootCid, true, row, col, dag)
-		}
-	}
-	// wait until enough data has been collected, too many errors encountered,
-	// or the timeout is reached
-	err := sc.wait()
-	if err != nil {
+	chunk := make([][]byte, edsWidth/2)
+	copy(chunk, rowRoots[:edsWidth/2])
+
+	errGroup, ctx := errgroup.WithContext(parentCtx)
+	errGroup.Go(func() error {
+		return fillQuarter(ctx, chunk, dag, true, dataSquare)
+	})
+	if err := errGroup.Wait(); err != nil {
 		return nil, err
 	}
-	// flatten the square
-	flattened := sc.flatten()
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
 
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
+	batchAdder := NewNmtNodeAdder(parentCtx, ipld.NewBatch(parentCtx, dag, ipld.MaxSizeBatchOption(batchSize(edsWidth))))
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
+	if err != nil {
+		return nil, err
 	}
-	return out
-}
 
-type index struct {
-	row uint32
-	col uint32
+	return extended, batchAdder.Commit()
 }
 
-type indexedShare struct {
-	data []byte
-	index
-}
-
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
-
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
-}
-
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuarter fetches 1/4 of shares for the given root
+func fillQuarter(
```

Comment:
Nit, but I would prefer `Quadrant` instead of `Quarter`

Reply:
Agree with John

Reply:
Renamed